libcudacxx/.upstream-tests/utils/nvidia/nvrtc/nvrtcc.cpp

```diff
@@ -0,0 +1,302 @@
+            [](const std::smatch& match) {
+                inputFile = match[0].str();
+                return NORMAL;
+            }
+    },
+    {
+        // If an input lists a .gpu file, run that file instead
+        std::regex("^([^-].*).gpu$"),
+            [](const std::smatch& match) {
+                execute = true;
+                executionConfig = ExecutionConfig {
+                    RunConfig{1, 0},
+                    {match[0].str()}
+                };
+
+                return NORMAL;
+            }
+    },
+    {
+        // If an input is a .exe file, search for other builds and run those
+        std::regex("^([^-].*).exe$"),
+            [](const std::smatch& match) {
+                execute = true;
+                executionConfig = load_execution_config_from_file(match[1].str() + ".build.yml");
+                assert(executionConfig.builds.size());
+                return NORMAL;
+            }
+    },
+    {
+        // Capture any argument not starting with '-' as the input file
+        std::regex("^([^-].+)[\\\\/].+$"),
+            [](const std::smatch& match) {
+                inputFile = match[0].str();
+                // Capture directory of input file as an include path
+                nvrtcArguments.emplace_back("-I " + match[1].str());
+                return NORMAL;
+            }
+    },
+    {
+        // Throw away remaining arguments
+        std::regex("^-.+$"),
+            [](const std::smatch& match) {
+                ignoredArguments.emplace_back(match[0].str());
+                return NORMAL;
+            }
+    },
+};
+
+int main(int argc, char **argv) {
+    // Greedily take off first arg
+    g_argc = argc-1;
+    g_argv = argv+1;
+
+    ArgProcessorState argState = NORMAL;
+
+    // Start by parsing arguments and building the configuration
+    std::string c_arg{};
+    for (auto a = g_argv; a < g_argv + g_argc; a++) {
+        // If the argument was greedy, we'll retry with an appended argument
+        c_arg = (argState == GREEDY) ? c_arg + " " + *a : *a;
+
+        for (auto& h: argHandlers) {
+            auto& regex = h.first;
+            auto& handler = h.second;
+
+            std::smatch matches;
+            std::regex_match(c_arg, matches, regex);
+
+            if (matches.size()) {
+                argState = handler(matches);
+                break;
+            }
+        }
+    }
+
+    fprintf(stderr, "NVRTCC Configuration:\r\n");
+    fprintf(stderr, "  Output dir: %s\r\n", outputDir.c_str());
+    fprintf(stderr, "  Input file: %s\r\n", inputFile.c_str());
+    fprintf(stderr, "  Building: %s\r\n", building ? "true" : "false");
+    fprintf(stderr, "  Executing: %s\r\n", execute ? "true" : "false");
+
+    // Load the input file and execute
+    if (execute) {
+        fprintf(stderr, "Executing %s with %i threads\r\n", executionConfig.builds[0].c_str(), executionConfig.rc.threadCount);
+        load_and_run_gpu_code(executionConfig.builds[0], executionConfig.rc);
```

Comment:
question: Instead of needing to pass the produced fatbin to `nvrtcc` in order to execute the program, couldn't `nvrtcc` just produce a binary that does this step for you?

Basically, it would store the fatbin in the binary, and in the `main()` it would do `load_and_run_gpu_code(/*stored fatbin*/)`.

That way, nvrtcc would work even more like a regular compiler in that it outputs a normal binary that can be executed just like `./a.out`.

Reply:
This is essentially what the old script did. It produced a binary with  the raw test stored inside and compiled it during execution. Every test was recompiled into the same hosting application. It bought us the 'works like a binary thing' but failed for giving us negative testing with .fail.cpp tests.

It might be possible to do this. It would get rid of needing to use the `PrefixExecutor()` for tests. I'll make a PR after this one with the documentation in mind.